src/kernels/MIOpenPReLU.cpp

```diff
@@ -0,0 +1,142 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+#include "tensor_view.hpp"
+
+template <typename TI, typename TO, unsigned NDIMS>
+__device__ void PReLUSWBackward(const TI* __restrict__ input,
+                                const TI* __restrict__ weight,
+                                const TO* __restrict__ output_grad,
+                                TI* __restrict__ input_grad,
+                                FLOAT_ACCUM* __restrict__ weight_grad_collector,
+                                ulong N,
+                                tensor_view_t<NDIMS> input_tv,
+                                tensor_view_t<NDIMS> output_grad_tv,
+                                tensor_view_t<NDIMS> input_grad_tv)
+{
+    int gid = blockIdx.x * blockDim.x + threadIdx.x;
+    if(gid >= N)
+        return;
+
+    auto tensor_layout  = tensor_layout_t<NDIMS>(input_tv, gid);
+    FLOAT_ACCUM input_v = CVT_FLOAT2ACCUM(input[input_tv.get_tensor_view_idx(tensor_layout)]);
```

Comment:
The index was calculated through tensor_layout rather than gid. Is this because there was no performance difference from the method of calculating gid as index assuming contiguous?

Reply:
I think the performance of a contiguous kernel would be better than this kernel. But as the performance have been significantly improved over ROCM in this implementation, I haven't implemented the contiguous kernel. By the way, in Modnn, this operation only have 5D kernels.

Reply:
You're right. moDNN also worked without assuming contiguous.